src/node/txreconciliation.cpp

```diff
@@ -363,79 +363,75 @@ class TxReconciliationTracker::Impl
                 std::holds_alternative<TxReconciliationState>(recon_state->second));
     }
 
-    bool IsFanoutTarget(const Wtxid& wtxid, NodeId peer_id, bool we_initiate, double n) const EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
+    std::tuple<std::vector<NodeId>, std::vector<NodeId>> GetFanoutTargets(const Wtxid& wtxid, size_t inbounds_fanout_tx_relay, size_t outbounds_fanout_tx_relay) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
     {
-        // We use the pre-determined randomness to give a consistent result per transaction, thus making sure that no transaction
-        // gets "unlucky" if every per-peer roll fails. This means that, given a `wtxid`, the ordering will always be the same,
-        // independently of what peer this is queried for. So some need to be picked eventually (as long as `n` doesn't shrink).
-        CSipHasher deterministic_randomizer{m_deterministic_randomizer};
-        deterministic_randomizer.Write(wtxid.ToUint256());
-
-        // We decide a peer is a fanout target for a given transaction deterministically at random based on two things: the wtxid of
-        // the transaction to be relayed (which we used as a seed for the randomizer) and the likelihood to fanout this transaction
-        // (which depends on how many non-erlay tx-relay peers we have, and the direction of the connection).
-        // In order to do this, we sort the peers in random order, and then we pick the top `n` peers of the resulting collection.
-        // If our chosen peer happens to be within the picked selection, we fanout to it, otherwise we reconcile.
-        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
-
-        std::vector<std::pair<uint64_t, NodeId>> best_peers;
-        best_peers.reserve(m_states.size());
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
 
+        // We decide whether a particular peer is a low-fanout flood target differently based on its connection direction:
+        // - for outbounds we have a fixed number of flood destinations.
+        // - for inbounds we use a fraction of all inbound peers supporting tx relay.
+        auto outbounds_target_size = (double)OUTBOUND_FANOUT_DESTINATIONS - outbounds_fanout_tx_relay;
+
+        // Since we use the fraction for inbound peers, we first need to compute the total number of inbound targets.
+        const double inbound_targets = (inbounds_fanout_tx_relay + m_inbounds_count) * INBOUND_FANOUT_DESTINATIONS_FRACTION;
+        double n = std::max(inbound_targets - inbounds_fanout_tx_relay, 0.0);
+
+        // Being this a fraction, we need to round it either up or down. We do this deterministically at random based on the
+        // transaction we are picking the peers for.
+        CSipHasher deterministic_randomizer_in{m_deterministic_randomizer};
```

Comment:
4234f5ebe133b949080aaa56da8cbdd18b650ff2

Originally, the code forced me to make this call for every peer, that's why i needed it to be deterministic. It's not the case anymore. Do we still need determinism then?

Reply:
Otherwise we can drop `m_deterministic_randomizer` altogether and use some cheap `FastRandomContext` instead.

Reply:
Well, this really depends. `GetFanoutTargets` is called by `RelayTransaction`, which, in turn, may be called by `ProcessMessage` for a peer that has `NetPermissionFlags::ForceRelay`. So if force relay is used for both fanout and set reconciliation, as suggested in https://github.com/bitcoin/bitcoin/pull/30116#discussion_r1818575030, then this could be called twice leading to two different sorting if we don't do so deterministically

Reply:
This could be `ReattemptInitialBroadcast`, `BroadcastTransaction` (rpc/wallet), `ProcessValidTx`, or earlier upon handling `NetMsgType::TX`.

I think it's odd we have two calls to `RelayTransaction` in the two latter cases. I'd rather change this code, e.g.  pass `already_relayed` (duplicating force) to `RelayTransaction`. Perhaps too much for this PR and we should do it elsewhere (especially now that the package stuff happens in-between). I understand why it was ok when `RelayTransaction` was simple.

And while we're stuck with these two calls, let's say we drop the determinism.

Why `RelayTransaction` doesn't look at `m_tx_inventory_to_send` very early (and maybe `m_tx_inventory_known_filter` too)? That would catch the transactions we already set for flooding, and just pass on them.

Now, the transactions in the sets. If we add them to the set again, nothing is done. If we try to flood them by a different choice, nothing is done still because we hit `m_tx_inventory_to_send` already having it.
 
Thus, for now, i think we should repeat the `m_tx_inventory_to_send` check above, rather than having the determinism. It kinda makes sense anyway, no?

Reply:
> Why RelayTransaction doesn't look at m_tx_inventory_to_send very early (and maybe m_tx_inventory_known_filter too)? That would catch the transactions we already set for flooding, and just pass on them.

I don't think this really works(?). 

Imagine we call this once, and peer `p` (who has ForceRelay` permissions) is selected for reconciliation using a non-deterministic approach.

Now, the transaction is sent again by `p` shortly after, and now he is selected for flooding instead. The filter won't catch this, and the transaction would end up in both sets.

This makes me recall that even though the process of selecting peers is deterministic at the moment, it is dependent on our peer list. Calling this twice with a different set of peers can result in a different ordering, so maybe we should not assume that transactions can be freely added to any of the two sets, and we should make sure that the opposite set does not contain the transaction before adding it to our desired one?